src/audio/copier.c

```diff
@@ -246,7 +246,11 @@ static struct comp_dev *create_dai(struct comp_dev *parent_dev, struct copier_da
 		dai.type = SOF_DAI_INTEL_ALH;
 		dai.is_config_blob = true;
 		type = ipc4_gtw_alh;
-		dai.dai_index -= IPC4_ALH_DAI_INDEX_OFFSET;
+		/* copier id = (group id << 4) + codec id + IPC4_ALH_DAI_INDEX_OFFSET
+		 * dai_index = (group id << 8) + codec id;
+		 */
+		dai.dai_index = ((dai.dai_index & 0xF0) << DAI_NUM_ALH_BI_DIR_LINKS_GROUP) +
+			(dai.dai_index & 0xF) - IPC4_ALH_DAI_INDEX_OFFSET;
```

Comment:
I fail to see how the comments are aligned with the code @RanderWang. You didn't make it easy on the reviewer or future user of this code :-)

specifically the minus sign in the end looks puzzling at best! @lgirdwood FYI.

Reply:
@plbossart dai index is 0-based, but codec id (not sdw spec defined) is calculated by windows driver according to following table copied from sof fw.    For example codec is on link0,   the group id = 0, codec id = 7, it will use dai 0

static const uint8_t alh_handshake_map[64] = {
	-1,	/* 0  - INVALID */
	-1,	/* 1  - INVALID */
	-1,	/* 2  - INVALID */
	-1,	/* 3  - INVALID */
	-1,	/* 4  - INVALID */
	-1,	/* 5  - INVALID */
	-1,	/* 6  - INVALID */
	22,	/* 7  - BIDIRECTIONAL */
	23,	/* 8  - BIDIRECTIONAL */
	24,	/* 9  - BIDIRECTIONAL */
	25,	/* 10 - BIDIRECTIONAL */
@bardliao please correct me

Reply:
group id = link id, codec id is 0-based.  the node id in blob is (group id << 4)  + codec id + 7. Current dai index in sof is group id << 8 + codec id.  why sof set group id << 8 ?  << 4 is enough and left 4 bits for codec id which can't be more than 16

Reply:
in our sdw linux driver , we also have a offset for pdi,  IPC4_ALH_DAI_INDEX_OFFSET  is kind of this offset. we usually use pdi->num = 2  as first normal pcm stream,  so value will be 2 + 2 + 3 = 7 + 0
````
	/* the Bulk and PCM streams are not contiguous */
	pdi->intel_alh_id = (link_id * 16) + pdi->num + 3;
	if (pdi->num >= 2)
		pdi->intel_alh_id += 2;
````

Reply:
close source FW unify the alh_id and dai index , they use the same algorithm   (x << 4) + y + 7 , for sof fw, we use this in driver but in fw we use (x << 8) + y as dai index.

Reply:
you missed the point @RanderWang 
dai_index = (group id << 8) + codec id; is straightforward
but you subtracted IPC4_ALH_DAI_INDEX_OFFSET without any explanations.
In the linux driver we ADD the offset to avoid using the first two PDIs that are for control. Here you subtract it.

I really don't get the algebra here.